### PR TITLE
add BrainFloat16 "float with less precision"

### PR DESCRIPTION
**Pull request by arnej27959**
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

@havardpe please review
@toregge @baldersheim @vekterli @geirst FYI

### CODE REVIEW

File: vespalib/src/vespa/vespalib/util/brain_float16.h [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+#pragma once
+
+#include <bit>
+#include <cstdint>
+#include <cstring>
+
+namespace vespalib {
+
+class BrainFloat16 {
```
Comment (baldersheim) [inline vespalib/src/vespa/vespalib/util/brain_float16.h:11]: Perhaps name it as the better known BFloat16 as that seems to be the more widely used name.

Comment (geirst) [reply vespalib/src/vespa/vespalib/util/brain_float16.h:11]: Yes, I agree. I guess the file names should reflect this as well.

Comment (geirst) [reply vespalib/src/vespa/vespalib/util/brain_float16.h:11]: Please also add class description. Could link to https://en.wikipedia.org/wiki/Bfloat16_floating-point_format as well.

File: vespalib/src/vespa/vespalib/util/brain_float16.h [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    static constexpr uint16_t float_to_bits(float value) noexcept {
+        TwoU16 both{0,0};
+        static_assert(sizeof(TwoU16) == sizeof(float));
+        memcpy(&both, &value, sizeof(float));
+        if constexpr (std::endian::native == std::endian::big) {
+            return both.u1;
+        } else if constexpr (std::endian::native == std::endian::little) {
+            return both.u2;
+        } else {
+            return 0;
```
Comment (havardpe) [inline vespalib/src/vespa/vespalib/util/brain_float16.h:42]: consider static assert with error message here

File: vespalib/src/vespa/vespalib/util/brain_float16.h [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    static constexpr float bits_to_float(uint16_t bits) noexcept {
+        TwoU16 both{0,0};
+        if constexpr (std::endian::native == std::endian::big) {
+            both.u1 = bits;
+        } else if constexpr (std::endian::native == std::endian::little) {
+            both.u2 = bits;
+        } else {
+            return 0.0;
```
Comment (havardpe) [inline vespalib/src/vespa/vespalib/util/brain_float16.h:53]: consider static assert

File: vespalib/src/tests/util/brain_float16/brain_float16_test.cpp [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+#include <vespa/vespalib/util/brain_float16.h>
+#include <vespa/vespalib/gtest/gtest.h>
+#include <stdio.h>
+#include <cmath>
+
+using namespace vespalib;
+
+using Limits = std::numeric_limits<BrainFloat16>;
+
```
Comment (havardpe) [inline vespalib/src/tests/util/brain_float16/brain_float16_test.cpp:11]: consider adding a test showcasing basic usage and (auto) conversion rules

File: vespalib/src/vespa/vespalib/util/brain_float16.h (hunk 1 above)
Comment (arnej27959) [reply vespalib/src/vespa/vespalib/util/brain_float16.h]: fixed.

File: vespalib/src/vespa/vespalib/util/brain_float16.h (hunk 2 above)
Comment (arnej27959) [reply vespalib/src/vespa/vespalib/util/brain_float16.h]: rewritten to allow static assert here and below.

File: vespalib/src/tests/util/brain_float16/brain_float16_test.cpp (hunk 4 above)
Comment (arnej27959) [reply vespalib/src/tests/util/brain_float16/brain_float16_test.cpp]: added test with normal values and showing how precision loss happen.

**Comment by arnej27959** [comment]
All comments should be considered now and more testing added, thank you all.

**Pull request merged**